Framework/Algorithms/src/FindPeaksConvolve.cpp

```diff
@@ -0,0 +1,376 @@
+
+  const Tensor1D smoothKernel{
+      createSmoothKernel(static_cast<size_t>(std::ceil(static_cast<double>(kernelBinCount) / 2.0)))};
+  const Tensor1D iOverSigConvOutput{(yConvOutput / eConvOutput).convolve(smoothKernel, dims)};
+
+  std::unique_ptr<EigenMap_const> xDataPostConv;
+  Eigen::VectorXd xDataCentredBins;
+  if (xData->size() != static_cast<size_t>(yData.size())) {
+    xDataCentredBins = centreBinsXData(xData);
+    xDataPostConv = std::make_unique<EigenMap_const>(EigenMap_const{xDataCentredBins.data(), xDataCentredBins.size()});
+  } else {
+    xDataPostConv = std::make_unique<EigenMap_const>(EigenMap_const(&xData->front(), xData->size()));
+  }
+  extractPeaks(dataIndex, iOverSigConvOutput, *xDataPostConv, yData, kernelBinCount / 2);
+
+  if (m_createIntermediateWorkspaces) {
+    createIntermediateWorkspaces(dataIndex, kernel, iOverSigConvOutput, *xDataPostConv);
+  }
+}
+
+Tensor1D FindPeaksConvolve::createKernel(const int binCount) {
+  Tensor1D kernel(binCount);
+  for (int i{0}; i < binCount; i++) {
+    // consider if these 0.25/0.75 values should be inputs
+    if (i < std::ceil(binCount) * 0.25 || i >= binCount * 0.75) {
+      kernel.data()[i] = -1.0;
+    } else {
+      kernel.data()[i] = 1.0;
+    }
+  }
+  return kernel;
+}
+
+Tensor1D FindPeaksConvolve::createSmoothKernel(const size_t kernelSize) {
+  Tensor1D kernel(kernelSize);
+  for (size_t i{0}; i < kernelSize; i++) {
+    kernel.data()[i] = 1.0 / static_cast<double>(kernelSize);
+  }
+  return kernel;
+}
+
+size_t FindPeaksConvolve::getKernelBinCount(const HistogramData::HistogramX *xData) const {
+  // What is the minimum number of data points this works for - add validation
+  const double peakExtent = getProperty("EstimatedPeakExtent");
+  const int peakExtentNBins = getProperty("EstimatedPeakExtentNBins");
+
+  if (peakExtent != EMPTY_DBL()) {
+    const double x1{xData->rawData()[static_cast<size_t>(std::floor((xData->size() - 1) / 2))]};
+    const double x2{xData->rawData()[static_cast<size_t>(std::floor((xData->size() - 1) / 2)) + 1]};
+    return static_cast<size_t>(std::floor(peakExtent * 2 / (x2 - x1)));
+  } else {
+    return static_cast<size_t>(peakExtentNBins);
+  }
+}
+
+Eigen::VectorXd FindPeaksConvolve::centreBinsXData(const HistogramData::HistogramX *xData) {
+  Eigen::VectorXd xDataVec{0.5 * (EigenMap_const(&xData->front(), xData->size() - 1) +
+                                  EigenMap_const(&xData->front() + 1, xData->size() - 1))};
+  return xDataVec;
+}
+
+void FindPeaksConvolve::extractPeaks(const size_t dataIndex, const Tensor1D &iOverSigma, const EigenMap_const &xData,
+                                     const TensorMap_const &yData, const size_t peakExtentBinNumber) {
+  auto filteredData = filterDataForSignificantPoints(iOverSigma);
+  int dataPointCount{0};
+  std::pair<int, double> dataRegionMax{0, 0.0};
+  std::vector<FindPeaksConvolve::PeakResult> peakCentres;
+  for (const auto &dataPoint : filteredData) {
```

Comment:
Here we do a second loop over data after we've already done a loop when copying the I/sigma array and zeroing some in `filterDataForSignificantPoints`. Could this not be done in a single loop?
For example when looping over I/sigma array, when you enter a peak region keep a track of the max value and corresponding index of either Y or I/sigma depending on `FindHighestDataPointInPeak` - if `True` then on leaving the peak region the position is just the max in Y, if `False` then we need to scale the y-values buy a normal pdf centered on the peak in I/sigma  - the latter will probably require additional loop but only over the elements in the peak region!?

Or perhaps if two loops in all cases then the first loop in `filterDataForSignificantPoints` should return `istart, iend` indices for each peak region found (see my comment below) and then the finding of the peak position need only loop over the elements between these indices.